endian/big.go

```diff
@@ -0,0 +1,8 @@
+//go:build mips || mips64 || ppc64 || s390x
+
+package endian
+
+import "encoding/binary"
+
+// Native is the platform's native byte order.
+var Native = binary.LittleEndian
```

Comment:
Ideally we can remove this package entirely and use https://github.com/josharian/native instead which also has runtime-based fallbacks for new GOARCH files in the future.

See activity in https://github.com/golang/go/issues/57237. I've been converting a number of Go ecosystem packages to using that.

@zx2c4, are you cool with that if we make you a co-maintainer of `josharian/native` for now?

cc @josharian

Reply:
I'm not really a fan of adding external dependencies, especially not ones that live in a random github repo. Yes, Josh is very far from "random" - I'd trust him to commit directly to wireguard-go need be - but it still means the general surface increases along a dimension that seems unnecessary.

But also, I'm not convinced we need this package here either. As far as I can tell, this is used in only this place:

```diff
+func (v *virtioNetHdr) decode(b []byte) error {
+	if len(b) < virtioNetHdrLen {
+		return errors.New("too short")
+	}
+	v.flags = b[0]
+	v.gsoType = b[1]
+	v.hdrLen = endian.Native.Uint16(b[2:])
+	v.gsoSize = endian.Native.Uint16(b[4:])
+	v.csumStart = endian.Native.Uint16(b[6:])
+	v.csumOffset = endian.Native.Uint16(b[8:])
+	return nil
+}
+
+func (v *virtioNetHdr) encode(b []byte) error {
+	if len(b) < virtioNetHdrLen {
+		return errors.New("too short")
+	}
+	b[0] = v.flags
+	b[1] = v.gsoType
+	endian.Native.PutUint16(b[2:], v.hdrLen)
+	endian.Native.PutUint16(b[4:], v.gsoSize)
+	endian.Native.PutUint16(b[6:], v.csumStart)
+	endian.Native.PutUint16(b[8:], v.csumOffset)
+	return nil
+}
```

If it's really just that, I can think of a few ways of handling this that might be simpler:

- Ensure that `b` is 2-byte aligned, check the length, and unsafe cast the buffer to a struct. This improves readability, IMO, since then you're filling out struct fields, rather than writing to given offsets.
- Don't ensure that `b` is 2-byte aligned, and just shrug shoulders at alignment traps on MIPS or ARM32 or whatever.
- Given that this struct appears to be only 10 bytes, populate it on the stack, and then memcpy it to the buffer, or vice versa. Hopefully the compiler is able to eliminate the extra loads and stores on x86 where alignment doesn't matter.
- Add a local private helper function for `u16`s, which does a little `if` statement like in Josh's `endian_generic.go`, and then rely on the compiler constant folding that to a no-op. Would be good to verify it does get folded away, and if not, find a different trick that does.

Any one of those solutions seems fine to me, and all are less bloaty than introducing a new package or relying on an external leftpad^Wnative package. Or some other creative solution. In other words, the fact that this doesn't actually _require_ a super generalized solution means we can relax our requirements a bit.

Then, long term, we can keep our fingers crossed about that native endian built-in Go package proposed last week.

Reply:
https://github.com/tailscale/wireguard-go/commit/439157bc5f01179180650cf3ebb949b5ecb38148 pulls in [github.com/josharian/native ](https://github.com/josharian/native) directly.

> Then, long term, we can keep our fingers crossed about that native endian built-in Go package proposed last week.

[github.com/josharian/native](https://github.com/josharian/native) is the closest thing to the proposal, and improves upon what we had previously (see `native/endian_generic.go`).

Reply:
I can't figure out how your reply is a complete response to what I wrote. Maybe I'm missing some subtle point of yours, but it seems like you just skipped over most of what I said.

Please do one of the four solutions I mentioned, or propose a fifth one, so long as its not adding a whole new package just to write out some `u16`s. Or, if you actually feel really strongly about adding a new package, you'll have to actually use words to make an argument that responds to what I wrote.

Reply:
By the way, the winning solution is likely "Given that this struct appears to be only 10 bytes, populate it on the stack, and then memcpy it to the buffer, or vice versa. Hopefully the compiler is able to eliminate the extra loads and stores on x86 where alignment doesn't matter.", by the way. Unless the Go compiler is really remarkably limited, this generally produces optimal code. Fill out a struct on the stack, memcpy it to the buffer, and hope the compiler coalesces the stores at native word size.

Reply:
> I can't figure out how your reply is a complete response to what I wrote. Maybe I'm missing some subtle point of yours, but it seems like you just skipped over most of what I said.

I was attempting to:
1. continue using the binary package over unsafe, with an eye towards https://github.com/golang/go/issues/57237 being implemented upstream
2. fix the case of new GOARCH doesn't build, with best effort assumption of little endian
3. avoid external deps by copying in github.com/josharian/native, with appropriate attribution

> By the way, the winning solution is likely "Given that this struct appears to be only 10 bytes, populate it on the stack, and then memcpy it to the buffer, or vice versa. Hopefully the compiler is able to eliminate the extra loads and stores on x86 where alignment doesn't matter.", by the way. Unless the Go compiler is really remarkably limited, this generally produces optimal code. Fill out a struct on the stack, memcpy it to the buffer, and hope the compiler coalesces the stores at native word size.

I implemented this approach in [798bed0](https://github.com/WireGuard/wireguard-go/pull/64/commits/798bed0b5d9283a45d9f3e78da7f5acd7477b7a6).